### ISSUE TITLE
[API Proposal]: Add API to enable BackgroundService startup to avoid blocking host startup

### ISSUE THREAD

**Issue by vova-lantsov-dev** [dotnet/runtime#36063]
## Background and motivation

IHostedService.StartAsync executes inline (and sequentially) when IHost.Start/StartAsync is called. With a BackgroundService, this causes lots of confusion since running code that isn't async in ExecuteAsync or StartAsync also runs inline. The workaround today is to call Task.Run on your own code or to use Task.Yield(), but this is not obvious.

## API Proposal

```diff
public class BackgroundService : IHostedService
{
+    public virtual bool StartAsynchronously { get; } = true;
}
```

NOTE: StartAsynchronously would be true by default.

## API Usage

```csharp
class MyBackgroundService : BackgroundService
{
    // It's true by default, this is just for illustration.
    public override bool StartAsynchronously => true;

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        using var timer = new PeriodicTimer(TimeSpan.FromSeconds(5));

        while (await timer.WaitForNextTickAsync(stoppingToken))
        {
            Console.WriteLine("Hello from background service!");
        }
    }
}
```

## Alternative Designs

An optional constructor argument.

<details>
<summary>Original Issue</summary>

BackgroundService blocked the execution of whole host

### Describe the bug
When I run the specific `foreach` cycle in `BackgroundService`-derived class, it blocks the whole host from starting. Even second hosted service doesn't start. When I comment `foreach` cycle, everything works as expected.

### To Reproduce
TargetFramework: netcoreapp2.1
Version: 2.1.12

Use following hosted service: [NotificationRunner.cs](https://gist.github.com/vova-lantsov-dev/7b3b820df1c7a6b86dac6bb0593c8155)
And singleton: [NotificationService.cs](https://gist.github.com/vova-lantsov-dev/788c5816a137a81e0c2af5f4655b21bc)

### Expected behavior
BackgroundService.ExecuteAsync should work in background without blocking even if it has blocking code. As you can see in [NotificationService](https://gist.github.com/vova-lantsov-dev/788c5816a137a81e0c2af5f4655b21bc) class, cancellation of enumerator is based on IApplicationLifetime.ApplicationStopping, but anyway it shouldn't affect the host startup because BackgroundService is expected to run in background :)

### Screenshots
When `foreach` cycle exists
![image](https://user-images.githubusercontent.com/45858087/62523649-ecafbf00-b83c-11e9-855e-e0ba541d0c3e.png)
Then execution is blocked on this cycle
![image](https://user-images.githubusercontent.com/45858087/62525002-4d3ffb80-b83f-11e9-8ce2-713dda7e79ad.png)

But when `foreach` cycle is commented
![image](https://user-images.githubusercontent.com/45858087/62525057-647ee900-b83f-11e9-84d0-47f896735b98.png)
Then execution continues as expected
![image](https://user-images.githubusercontent.com/45858087/62525220-ae67cf00-b83f-11e9-83a9-c7ca1c45b82c.png)
(But why twice?)

### Additional context
```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview7-012821
 Commit:    6348f1068a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview7-012821\

Host (useful for support):
  Version: 3.0.0-preview7-27912-14
  Commit:  4da6ee6450

.NET Core SDKs installed:
  2.1.700 [C:\Program Files\dotnet\sdk]
  2.1.701 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.2.300 [C:\Program Files\dotnet\sdk]
  2.2.301 [C:\Program Files\dotnet\sdk]
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.0.100-preview7-012821 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview7.19365.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview7-27912-14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview7-27912-14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

</details>

**Comment by HelloKitty** [comment]
I noticed that hangs can now occur in later versions of ASP Core if you do a Task.Delay. Previous versions of ASP Core were not causing this issue but I don't have exact version information. Calling `await Task.Delay(JobConfig.IntervalMilliseconds, cancellationToken);` within `public async Task StartAsync(CancellationToken cancellationToken)` In ASP Core 3.1 now, unlike a previous version, blocks startup for some reason.

**Comment by vova-lantsov-dev** [comment]
@HelloKitty `IHostedService.StartAsync` always blocks the execution of the whole host. This is expected behaviour. Try to use `BackgroundService.ExecuteAsync` for non-blocking delayed tasks.

**Comment by HelloKitty** [comment]
@vova-lantsov-dev It appears so, but I do not think this was the case in a previous version of ASP Core. This code did not stop startup execution in a previous version, I have adopted `BackgroundService.ExecuteAsync` and now delay in there.

**Comment by drewkill32** [comment]
This causes an issue when unit testing. If I mock a service to return a `CompletedTask`  instead of an awaited task `StartAsync` never returns. 

```csharp
 [TestClass]
public class TestClass
{
        [TestMethod, TestCategory("UnitTest")]
        public async Task Background_Service_Should_Return_FromStartAsync()
        {
            
            var sut = new FakeService();

            await sut.StartAsync(Token);

            //code never reaches here because ExecuteAsync never awaits a task
            Assert.IsTrue(true);
        }
}
public class FakeService : BackgroundService
{
        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                await Task.CompletedTask;
            }
        }
}

```

**Comment by BowserKingKoopa** [comment]
Just ran into this. Fixed with await Task.Yield() at top of BackgroundService.ExecuteAsync. I think this is especially confusing because we don't call ExecuteAsync ourselves, the framework does. We don't even instantiate the background service ourselves (we're using services.AddHostedService<MyBackgroundService>()). So the fact that the instantiation and call to ExecuteAsync happen as part of the ASP.NET app startup is not obvious. In fact it's downright strange imho.

**Comment by pierregaboriaud** [comment]
Or if it doesn't work you can use Task.Run

`
protected override Task ExecuteAsync(CancellationToken stoppingToken) => **Task.Run(async () =>**
{
    while (!stoppingToken.IsCancellationRequested)
    {
        // Do async work
        await Task.Delay(2000, stoppingToken);
    }
});
`

**Comment by maryamariyan** [comment]
Since a workaround exists for this issue and it is not a regression it is not a critical fix for 5.0, (moving to 6.0).

**Comment by roblapp** [comment]
I have a background process that has long periods of time that are blocking. What is the appropriate setup to have it run during the lifecycle of my application? Adding an `await Task.Delay` at the beginning of `ExecuteAsync`? Or using `Task.Run(() => {...});` in `ExecuteAsync`?

**Comment by vova-lantsov-dev** [comment]
@roblapp both should work for you

But you forgot about 'async':
`Task.Run(async () => ...)`

**Comment by GaryShao** [comment]
I use task run to avoid this issue. but it's a little bit confused, I'm told that do not use TASK to execute long running task. Because task is consuming threads from thread resource pool which is design for short running tasks.

**Comment by Assam** [comment]
Based on the above I think [documentation](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/multi-container-microservice-net-applications/background-tasks-with-ihostedservice) should be updated. It seems below class is doing blocking operation on ExecuteAsync
```
public class GracePeriodManagerService : BackgroundService
{
    private readonly ILogger<GracePeriodManagerService> _logger;
    private readonly OrderingBackgroundSettings _settings;

    private readonly IEventBus _eventBus;

    public GracePeriodManagerService(IOptions<OrderingBackgroundSettings> settings,
                                     IEventBus eventBus,
                                     ILogger<GracePeriodManagerService> logger)
    {
        // Constructor's parameters validations...
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        _logger.LogDebug($"GracePeriodManagerService is starting.");

        stoppingToken.Register(() =>
            _logger.LogDebug($" GracePeriod background task is stopping."));

        while (!stoppingToken.IsCancellationRequested)
        {
            _logger.LogDebug($"GracePeriod task doing background work.");

            // This eShopOnContainers method is querying a database table
            // and publishing events into the Event Bus (RabbitMQ / ServiceBus)
            CheckConfirmedGracePeriodOrders();

            await Task.Delay(_settings.CheckUpdateTime, stoppingToken);
        }

        _logger.LogDebug($"GracePeriod background task is stopping.");
    }

    .../...
}
```

**Comment by manuelelucchi** [comment]
I noticed this happens while implementing IHostedService manually but even if you use ``Task.Run`` or ``Task.Yield`` on the ``StartAsync`` it doesn't work. You need ``BackgroundService`` and ``ExecuteAsync``

**Comment by Goffen** [comment]
Wow what a wierd behaviour from a "background service"

**Comment by jonas-lomholdt** [comment]
I hid this issue yesterday. While implementing `BackgroundService` it worked without any `Task.Yield()`, but today it doesn't. Does anyone know how that that is possible? Adding the `await Task.Yield()` at the top fixed it, but I don't understand how it *could* work without it.

**Comment by rjgotten** [comment]
> While implementing `BackgroundService` it worked without any `Task.Yield()`, but today it doesn't. Does anyone know how that is possible? 

Look at [the implementation of `BackgroundService`](https://github.com/dotnet/runtime/blob/v6.0.3/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs).

`StartAsync` is a **synchronous** (!!) method which returns either the `Task` taken as result from `ExecuteAsync`, if said task `IsCompleted`; and otherwise returns a pre-completed `Task.CompletedTask`. It will take responsbility to stop the execute task in `StopAsync` by triggering the cancellation token, regardless.

Because `StartAsync` is synchronous, the code _will block_ until the actual `ExecuteAsync` implementation, if it is an async method, does its first `await`.  That first `await` is where the asynchronous state machine compiler magic effectively kicks in and the async method returns its `Task` in a yet-to-be-completed state.

**Comment by MartinDemberger** [comment]
Another problem with this behaviour is with exceptions.

When the service throws a exception in the first line the process is stopped even if `HostOptions.BackgroundServiceExceptionBehavior` is set to `BackgroundServiceExceptionBehavior.Ignore`.

```csharp
   public class ServiceWithError : BackgroundService
    {
        //protected override Task ExecuteAsync(CancellationToken stoppingToken) => Task.Run(() => throw new System.NotImplementedException());
        protected override Task ExecuteAsync(CancellationToken stoppingToken) => throw new System.NotImplementedException();
    }
```

**Comment by eerhardt** [comment]
This issue hasn't been planned for 7.0 (see https://github.com/dotnet/runtime/issues/64015). Moving to Future.

**Comment by suchoss** [comment]
> Async method executes synchronously till first `await` is reached (it means `BackgroundService.ExecuteAsync` won't return till `await` is called). So this problem is not related to ASP.NET Core, but to async compilation.

I would like to point out that the "executes synchronously till first `await` is reached" is not true in my case. 

```csharp
protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        using var timer = new PeriodicTimer(TimeSpan.FromHours(24));
        
        // If I do not add following line, the code stops for very long time. More than 1 hour.
        await Task.Yield();  
        
        do
        {
            logger.Debug($"Timer triggered.");
            var count = Interlocked.Increment(ref executionCount);
            logger.Debug($"Trigger counter:{count}.");
            
            // this await wont return control back to main program
            // it is still blocking
            await RunLoop(stoppingToken);
            
        } while (await timer.WaitForNextTickAsync(stoppingToken));
    }

    private async Task RunLoop(CancellationToken stoppingToken)
    {
        foreach (var indexType in Enum.GetValues<AutocompleteIndexType>())
        {
            logger.Debug($"Creating {indexType:G} index.");
            try
            {
                // this await wont return control back to main program
                // it is still blocking - probably because 90 % of following calls are "synchronous" inside
                await IndexFactory.CreateCacheIndex(indexType);
            }
            catch (Exception e)
            {
                logger.Error($"Error occured during {indexType:G} index creation.", e);
            }
        }
        
        logger.Debug("Running delete of old indexes.");
        try
        {
            IndexFactory.DeleteOld();

        }
        catch (Exception e)
        {
            logger.Error($"Problem occured when deleting old files.", e);
        }
    }
```

But `await Task.Yield();` at the beginning works.

**Comment by Leandro-Albano** [comment]
I just ran into this issue, and in my case it seems that `await Task.Yield();` doesn’t help. In my case, it also seem to be a problem just on Debug mode. I have the following:

public class ConsumerBackgroundService : BackgroundService
    {
        protected override async Task ExecuteAsync(CancellationToken cancellationToken)
        {
            await Task.Yield();
            while (!cancellationToken.IsCancellationRequested)
            {
            }
        }

        public override Task StopAsync(CancellationToken cancellationToken) => base.StopAsync(cancellationToken);

    }

**Comment by Leandro-Albano** [comment]
So, I tried the above code, and apparently the issue only happens on Mac, on both vscode and vs for Mac.

**Comment by thomrad** [comment]
> So, I tried the above code, and apparently the issue only happens on Mac, on both vscode and vs for Mac.

Yes, can confirm this message. Debug build in Rider on Mac blocks, Run in Rider on Mac works.

**Comment by montella1507** [comment]
We have simple Backround service like this:

`

    public class RefresherService : BackgroundService
    {


        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            try
            {

                while (!stoppingToken.IsCancellationRequested)
                {
                    try
                    {

                      await Do();
                    }
                    catch (Exception ex)
                    {
                      
                    }
                    finally
                    {
                   
                        await Task.Delay(TimeSpan.FromMinutes(1), stoppingToken);
                    }
                }
            }
            catch (Exception ex) when (ex is TaskCanceledException || ex is OperationCanceledException)
            {
           
            }
            catch (Exception ex)
            {
           
            }
        }
    
}`

And it B L O C K S the startup of the server for 1 minute.... This is not how should have  the BackgroundService and ExecuteAsync behave...

**Comment by Davilink** [comment]
Maybe we should have 2 class of background service:
`BlockingBackgroundService` and `NonBlockingBackgroudService`
`BlockingBackgroundService` would use the same current implement of backgroundService and the `NonBlockingBackgroudService` would spawn internally spawn a Task.Run() just like demonstrated in this commit : https://github.com/dotnet/extensions/commit/8d0b1284ea8985eef7d52c0044f78e7c2fc26052#diff-a2481caefbc8935408d8a841d3040f716cac00cbd846ef84c9892ce4046c59d9R37

**Comment by christallire** [comment]
I've run into this too. 
@eerhardt could we make this happen in 8.0? this could be a relatively simple fix compared to the time spent scratching their heads :)

**Comment by Xor-el** [comment]
I think this should make it in NET 8.0 as indicated here
https://github.com/dotnet/runtime/issues/86511
We have gotten two new properties added that should resolve this.

https://learn.microsoft.com/dotnet/api/microsoft.extensions.hosting.hostoptions.servicesstartconcurrently?view=dotnet-plat-ext-8.0

https://learn.microsoft.com/dotnet/api/microsoft.extensions.hosting.hostoptions.servicesstopconcurrently?view=dotnet-plat-ext-8.0

**Comment by StephenCleary** [comment]
The `IHostedLifecycleService` and `ServicesStartConcurrently` changes in .NET 8.0 do not address this issue. The host (as currently written) [still starts the services by calling `StartAsync` directly](https://github.com/dotnet/runtime/blob/1487d46e4ff7731b52af92b1228373a2a66ed6bf/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L133). So a synchronous implementation of `StartAsync` will still block host startup.

**Comment by naumenkoff** [comment]
[.NET 8 has been released](https://devblogs.microsoft.com/dotnet/announcing-dotnet-8/), and unfortunately, Microsoft has turned a blind eye to this issue. 
For those who have encountered it, it would be wise not to use BackgroundService. Instead, you can create your implementation based on [IHostedService](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-8.0) or the new [IHostedLifecycleService](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedlifecycleservice?view=dotnet-plat-ext-8.0).

The situation is amusing - to start a synchronous service, you need to await something, for example, Task.Yield, resulting in the generation of a [state machine](https://devblogs.microsoft.com/dotnet/how-async-await-really-works/) and a loss of at least 0.05% performance at this stage. 
Alternatively, you can use await Task.Run and lose at least 0.10% simply because of the presence of a [state machine](https://devblogs.microsoft.com/dotnet/how-async-await-really-works/) to start the service elegantly with just one addition to the DI container.

**Comment by montella1507** [comment]
Yeah.. MS is way more focusing on implementing YET next type of UI library, just to deprecate and kill it 1 year later, rather then solve these issues..

**Comment by davidfowl** [comment]
I’m very confused why Task.Run isn’t a viable workaround?

**Comment by naumenkoff** [comment]
> I’m very confused why Task.Run isn’t a viable workaround?

You can do that until the presence of a state machine becomes a hindrance to you, and the [TaskScheduler](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler?view=net-7.0) works perfectly.

**Comment by davidfowl** [comment]
> You can do that until the presence of a state machine becomes a hindrance to you

When does that happen?

**Comment by JohannSig** [comment]
This sounds like a "rite of passage" pitfall for anyone deciding to stroll into BackgroundService country. Myself included - twice now (2020 + 2024) 😆

**Comment by JohannSig** [comment]
Wonder where the knife lies in this cow?

```
    public ConcurrentQueue<SearchAvailabilitiesParameters> SearchesToStart { get; set; } = new();

    public ConcurrentDictionary<string, Task> SearchesRunning { get; set; } = new();

    public void EnqueueSearch(SearchAvailabilitiesParameters parameters)
    {
        SearchesToStart.Enqueue(parameters);
    }
    
    protected async override Task ExecuteAsync(CancellationToken stoppingToken)
    {
        await Task.Yield();

        try
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                await Task.Delay(50, stoppingToken);

                while (SearchesToStart.TryDequeue(out var parameters))
                {
                    var cacheKey = parameters.ToCacheKey();

                    SearchesRunning.TryAdd(cacheKey, StartSearchAsync(parameters, stoppingToken));
                }
            }
        }
        catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
        {
            return;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Exception while managing background searches.");

            return;
        }
    }
```

**Comment by jesslilly** [comment]
### Nothing here was working for me...
🔴 As it turns out you **cannot test this with breakpoints!**  

Maybe that seems logical to some people, but it did not dawn on me until I spent an hour, tried every single solution and workaround in this issue, and finally said, to myself, "Hmmm, maybe I'll try logging instead of using a breakpoint".  

Give me a thumbs up if this was helpful.  And give me a pie in the face otherwise!  🤣

**Comment by BernhardPollerspoeck** [comment]
> ... Most developers don't think about `async` methods as starting synchronously. There's a number of SO questions that are variants of "I started with a background service template and it works fine until I try to remove the Task.Delay" or "it stops working as soon as I try to [synchronously] read from a queue of work". For those of us with a good understanding of async _and_ how the host starts background services, the problem is obvious; but that's not most developers.

this sounds more like a edjucation/knownledge or documentation issue then. Just slapping async/await onto everything is not "working with tasks". 
Maybe a analyzer would help a lot here to clarify and guide?

**Comment by christallire** [comment]
Everything should've been built in a foolproof way.

**Comment by Bertramp** [comment]
Just hit this issue _again_ because some of our senior devs had never used .NET background services and didn't know about this caveat. I assumed it had been fixed while reviewing their PR's since they are very experienced. Turns out they too just fell into the newb trap.

Something called `BackgroundService` simply _cannot_ under any circumstance be allowed to _not_ run on a background thread. Currently the default behaviour of `BackgroundService` just hides the fact that `StartAsync` simply calls `ExecuteAsync`. Any sane person would assume that since `StartAsync` is not declared abstract, it's because it handles dispatching the actual work to a background thread, and that you shouldn't mess with that implementation unless you really have to.

**Comment by ericstj** [comment]
I think that's a reasonable breaking change to take.   Is a property or constructor parameter on `BackgroundService` sufficient for the control of this?  I imagine in most cases the derived class would chaose one option or the other.  Anyone want to tackle converting this to an API proposal?  cc @steveharter

**Comment by BowserKingKoopa** [comment]
This is such a funny issue. On the one hand you'd think starting your "Background Service" would run in the background. On the other hand async/await runs sync until it hits something async. So you either get a weird idea of an 'eventually background service' or you make this a special case where await doesn't actually do what await does everywhere else in .net. Is this just fallout from the weird 'runs sync until async' thing. That can bite you in other circumstances too.  It's weird, or maybe 'unexpected' or 'unintuitive' rather.

For my background services I'm just Task.Yielding at the start of them reflexively now so I don't even have to think about when it switches to actually async.

**Comment by davidfowl** [comment]
> I think that's a reasonable breaking change to take. Is a property or constructor parameter on BackgroundService sufficient for the control of this? I imagine in most cases the derived class would chaose one option or the other. Anyone want to tackle converting this to an API proposal? cc @steveharter

Yep! I took a stab at it, and marked it ready for review. Feel free to edit the proposal or suggest better naming (or patterns).

**Comment by terrajobst** [comment]
* We don't believe many people will be broken by the new behavior
    - People who are, can either implement `IHostedService` or do their own dependency handling
* IOW, we like the behavior breaking change but not the API.
    - We will close this issue (as it tracks the API). There will be another issue.

```C#
namespace Microsoft.Extensions.Hosting;

public class BackgroundService : IHostedService
{
    public virtual bool StartAsynchronously { get; } = true;
}
```

**Comment by StephenCleary** [comment]
I like and agree with just making this always on. Can you link the new issue number here when available?

**Comment by davidfowl** [comment]
Only took us 5 years 😄

**Comment by davidfowl** [comment]
> We will close this issue (as it tracks the API). There will be another issue. 

Did we do this?

**Comment by ericstj** [comment]
I can't find evidence of this happening, and couldn't find any issue opened by either @terrajobst or @steveharter.  Let's just use this issue.
